### ISSUE TITLE
`dapr/components-contrib` and `dapr/dapr` interdependence problem.

### ISSUE THREAD

**Issue by 1046102779** [dapr/dapr#5085]
https://github.com/dapr/dapr/blob/master/go.mod#L560
```go.mod
// TEMP ONLY
replace github.com/dapr/components-contrib => github.com/italypaleale/dapr-components-contrib v1.4.0-rc1.0.20220822232050-9ace33ef3b7f
```

This approach is not an effective way in the long run. If master temporarily uses github's own account branch, and then branches from other children's shoes account, it is different from this branch, which will lead to failure

```shell
# github.com/dapr/dapr/pkg/grpc
../../pkg/grpc/api.go:1586:16: v.Key undefined (type *configuration.Item has no field or method Key)
../../pkg/grpc/api.go:1610:16: v.Key undefined (type *configuration.Item has no field or method Key)
# github.com/dapr/dapr/pkg/http
../../pkg/http/api.go:801:96: item.Key undefined (type *configuration.Item has no field or method Key)
../../pkg/http/api.go:950:47: item.Key undefined (type *configuration.Item has no field or method Key)
```

Then I feel that  the follow is another problem:

`dapr/components-contrib` and `dapr/dapr` have dependencies on each other.

1. For `dapr/components-contrib` repo depends on `dapr/dapr` repo. 
```shell
➜  components-contrib git:(feature/secretstores) grep -rn "github.com/dapr/dapr" . |grep go.mod
./tests/certification/bindings/rabbitmq/go.mod:8:       github.com/dapr/dapr v1.8.4-0.20220822154328-9797b4006e23
./tests/certification/bindings/rabbitmq/go.mod:134:replace github.com/dapr/dapr => github.com/berndverst/dapr v1.1.3-0.20220825205949-5d389ea9a3bb
./tests/certification/bindings/rabbitmq/go.mod:141:// replace github.com/dapr/dapr => ../../../../../dapr
./tests/certification/bindings/azure/eventhubs/go.mod:8:        github.com/dapr/dapr v1.8.4-0.20220822154328-9797b4006e23
./tests/certification/bindings/azure/eventhubs/go.mod:163:replace github.com/dapr/dapr => github.com/berndverst/dapr v1.1.3-0.20220825205949-5d389ea9a3bb
./tests/certification/bindings/azure/cosmosdb/go.mod:9: github.com/dapr/dapr v1.8.4-0.20220822154328-9797b4006e23
./tests/certification/bindings/azure/cosmosdb/go.mod:156:replace github.com/dapr/dapr => github.com/berndverst/dapr v1.1.3-0.20220825205949-5d389ea9a3bb
./tests/certification/bindings/azure/blobstorage/go.mod:9:      github.com/dapr/dapr v1.8.4-0.20220822154328-9797b4006e23
./tests/certification/bindings/azure/blobstorage/go.mod:154:replace github.com/dapr/dapr => github.com/berndverst/dapr v1.1.3-0.20220825205949-5d389ea9a3bb
./tests/certification/bindings/azure/servicebusqueues/go.mod:8: github.com/dapr/dapr v1.8.4-0.20220822154328-9797b4006e23
./tests/certification/bindings/azure/servicebusqueues/go.mod:160:replace github.com/dapr/dapr => github.com/berndverst/dapr v1.1.3-0.20220825205949-5d389ea9a3bb
./tests/certification/bindings/azure/storagequeues/go.mod:8:    github.com/dapr/dapr v1.8.4-0.20220822154328-9797b4006e23
./tests/certification/bindings/azure/storagequeues/go.mod:156:replace github.com/dapr/dapr => github.com/berndverst/dapr v1.1.3-0.20220825205949-5d389ea9a3bb
./tests/certification/bindings/kafka/go.mod:10: github.com/dapr/dapr v1.8.4-0.20220822154328-9797b4006e23
./tests/certification/bindings/kafka/go.mod:151:replace github.com/dapr/dapr => github.com/berndverst/dapr v1.1.3-0.20220825205949-5d389ea9a3bb
./tests/certification/bindings/kafka/go.mod:158:// replace github.com/dapr/dapr => ../../../../../dapr
./tests/certification/bindings/alicloud/nacos/go.mod:8: github.com/dapr/dapr v1.8.4-0.20220822154328-9797b4006e23
./tests/certification/bindings/alicloud/nacos/go.mod:140:replace github.com/dapr/dapr => github.com/berndverst/dapr v1.1.3-0.20220825205949-5d389ea9a3bb
./tests/certification/bindings/alicloud/dubbo/go.mod:10:        github.com/dapr/dapr v1.8.4-0.20220822154328-9797b4006e23
./tests/certification/bindings/alicloud/dubbo/go.mod:167:replace github.com/dapr/dapr => github.com/berndverst/dapr v1.1.3-0.20220825205949-5d389ea9a3bb
./tests/certification/pubsub/rabbitmq/go.mod:9: github.com/dapr/dapr v1.8.4-0.20220822154328-9797b4006e23
./tests/certification/pubsub/rabbitmq/go.mod:134:replace github.com/dapr/dapr => github.com/berndverst/dapr v1.1.3-0.20220825205949-5d389ea9a3bb
./tests/certification/pubsub/rabbitmq/go.mod:141:// replace github.com/dapr/dapr => ../../../../../dapr
./tests/certification/pubsub/azure/servicebus/go.mod:8: github.com/dapr/dapr v1.8.4-0.20220822154328-9797b4006e23
./tests/certification/pubsub/azure/servicebus/go.mod:160:replace github.com/dapr/dapr => github.com/berndverst/dapr v1.1.3-0.20220825205949-5d389ea9a3bb
./tests/certification/pubsub/azure/eventhubs/go.mod:8:  github.com/dapr/dapr v1.8.4-0.20220822154328-9797b4006e23
./tests/certification/pubsub/azure/eventhubs/go.mod:162:replace github.com/dapr/dapr => github.com/berndverst/dapr v1.1.3-0.20220825205949-5d389ea9a3bb
./tests/certification/pubsub/azure/eventhubs/go.mod:169:// replace github.com/dapr/dapr => ../../../../../dapr
./tests/certification/pubsub/kafka/go.mod:10:   github.com/dapr/dapr v1.8.4-0.20220822154328-9797b4006e23
./tests/certification/pubsub/kafka/go.mod:151:replace github.com/dapr/dapr => github.com/berndverst/dapr v1.1.3-0.20220825205949-5d389ea9a3bb
./tests/certification/pubsub/kafka/go.mod:158:// replace github.com/dapr/dapr => ../../../../../dapr
./tests/certification/pubsub/mqtt/go.mod:9:     github.com/dapr/dapr v1.8.4-0.20220822154328-9797b4006e23
./tests/certification/pubsub/mqtt/go.mod:140:replace github.com/dapr/dapr => github.com/berndverst/dapr v1.1.3-0.20220825205949-5d389ea9a3bb
./tests/certification/pubsub/mqtt/go.mod:147:// replace github.com/dapr/dapr => ../../../../../dapr
./tests/certification/state/azure/blobstorage/go.mod:8: github.com/dapr/dapr v1.8.4-0.20220822154328-9797b4006e23
./tests/certification/state/azure/blobstorage/go.mod:154:replace github.com/dapr/dapr => github.com/berndverst/dapr v1.1.3-0.20220825205949-5d389ea9a3bb
./tests/certification/state/azure/tablestorage/go.mod:8:        github.com/dapr/dapr v1.8.4-0.20220822154328-9797b4006e23
./tests/certification/state/azure/tablestorage/go.mod:154:replace github.com/dapr/dapr => github.com/berndverst/dapr v1.1.3-0.20220825205949-5d389ea9a3bb
./tests/certification/state/sqlserver/go.mod:8: github.com/dapr/dapr v1.8.4-0.20220822154328-9797b4006e23
./tests/certification/state/sqlserver/go.mod:135:replace github.com/dapr/dapr => github.com/berndverst/dapr v1.1.3-0.20220825205949-5d389ea9a3bb
./tests/certification/state/redis/go.mod:8:     github.com/dapr/dapr v1.8.4-0.20220822154328-9797b4006e23
./tests/certification/state/redis/go.mod:137:replace github.com/dapr/dapr => github.com/berndverst/dapr v1.1.3-0.20220825205949-5d389ea9a3bb
./tests/certification/state/postgresql/go.mod:8:        github.com/dapr/dapr v1.8.4-0.20220822154328-9797b4006e23
./tests/certification/state/postgresql/go.mod:140:replace github.com/dapr/dapr => github.com/berndverst/dapr v1.1.3-0.20220825205949-5d389ea9a3bb
./tests/certification/secretstores/azure/keyvault/go.mod:8:     github.com/dapr/dapr v1.8.4-0.20220822154328-9797b4006e23
./tests/certification/secretstores/azure/keyvault/go.mod:155:replace github.com/dapr/dapr => github.com/berndverst/dapr v1.1.3-0.20220825205949-5d389ea9a3bb
./tests/certification/secretstores/local/file/go.mod:8: github.com/dapr/dapr v1.8.4-0.20220822154328-9797b4006e23
./tests/certification/secretstores/local/file/go.mod:130:replace github.com/dapr/dapr => github.com/berndverst/dapr v1.1.3-0.20220825205949-5d389ea9a3bb
./tests/certification/secretstores/local/env/go.mod:8:  github.com/dapr/dapr v1.8.4-0.20220822154328-9797b4006e23
./tests/certification/secretstores/local/env/go.mod:130:replace github.com/dapr/dapr => github.com/berndverst/dapr v1.1.3-0.20220825205949-5d389ea9a3bb
./tests/certification/go.mod:7: github.com/dapr/dapr v1.8.4-0.20220822154328-9797b4006e23
./tests/certification/go.mod:132:replace github.com/dapr/dapr => github.com/berndverst/dapr v1.1.3-0.20220825205949-5d389ea9a3bb
```

2. For `dapr/dapr` repo depends on `dapr/components-contrib` repo. 

The problem is that we are dealing with incompatible problems. For example, the API context PR(https://github.com/dapr/components-contrib/pull/1973) is a breaking change, so I don't know how to deal with it, and at the same time I encounter the first problem `../../pkg/grpc/api.go:1586:16: v.Key undefined` above. So I have to wait for the `go.mod` above to be changed to `dapr/components-contrib` before I can change my changes, which is queued and blocked here.

/cc @ItalyPaleAle @yaron2 @artursouza @daixiang0 .

I don't think there should be a fork library in master, and I don't know if there is a better way to deal with the current problems?

**Comment by ItalyPaleAle** [comment]
The fork is temporary and was meant to be there for a few hours tops. Due to a series of PRs that got piled up, we had to do some more work and right now both Dapr/dapr and components-contrib have pinned forks (from me and @berndverst ). This was an exceptional situation and we'll have it fixed by EOD Friday. Sorry for the inconvenience.

And yes, we know that the circular dependency is a problem. Unless we migrate to a mono repo (which would have its own issues) we will always have these brief, "awkward" situations sadly (and that's why I am not supportive of adding yet another repo like you suggest in #4923)

**Comment by 1046102779** [comment]
> The fork is temporary and was meant to be there for a few hours tops. Due to a series of PRs that got piled up, we had to do some more work and right now both Dapr/dapr and components-contrib have pinned forks (from me and @berndverst ). This was an exceptional situation and we'll have it fixed by EOD Friday. Sorry for the inconvenience.
> 
> And yes, we know that the circular dependency is a problem. Unless we migrate to a mono repo (which would have its own issues) we will always have these brief, "awkward" situations sadly (and that's why I am not supportive of adding yet another repo like you suggest in #4923)

https://github.com/dapr/components-contrib/pull/1973

I will encounter this later when I change the ·secretstores· API context destructively.

another problem:

```golang
./tests/certification/bindings/kafka/go.mod:151:replace github.com/dapr/dapr => github.com/berndverst/dapr v1.1.3-0.20220825205949-5d389ea9a3bb
```

This kind of replacement is really a bit uncomfortable, because I will do it too.

**Comment by 1046102779** [comment]
Compatibility must be considered for subsequent changes, which is really painful, because the speed of iteration and considerations will be very high, and it will be very cautious and daunting to do.

**Comment by berndverst** [comment]
It is unfortunate but was unavoidable here. I spent my entire day on this interdependence problem!

Once my PR #5083 is merged the issue should be resolved again for now.

In the future we may choose to break certification tests in contrib temporarily to avoid this interdependence issue.

Unfortunately it is not possible to make all tests pass and not having to create temporary forks like was done here. I agree that we should avoid doing so in dapr/dapr.

Bernd (maintainer of contrib)

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by daixiang0** [comment]
I agree that we should avoid doing so in dapr/dapr.

**Comment by 1046102779** [comment]
At present, if it is an API (`components-contrib` repo) destructive change, it seems that this is unavoidable. Can we introduce a branch of the branch to deal with this problem, such as: the `develop` branch is used to merge, if there is no problem, and then merge into the master and so on way.

/cc @daixiang0 @berndverst @ItalyPaleAle

**Comment by daixiang0** [comment]
I want to ask why you always sync master branch, is it for development?

**Comment by 1046102779** [comment]
> We will regularly synchronize the code of the community master branch to the internal, if we use replace, we cannot use replace internally, because there are differences in replace.

We will regularly synchronize the code of the community master branch to the internal, if we use replace, we cannot use replace internally, because there are differences in replace.

**Comment by ItalyPaleAle** [comment]
@1046102779 the replace has been removed now

**Comment by 1046102779** [comment]
> @1046102779 the replace has been removed now

Thanks.
